pkg/agent/servicecidr/discoverer.go

```diff
@@ -79,43 +119,48 @@ func (d *discoverer) GetServiceCIDR(isIPv6 bool) (*net.IPNet, error) {
 	return d.serviceIPv4CIDR, nil
 }
 
-func (d *discoverer) AddEventHandler(handler EventHandler) {
+func (d *Discoverer) AddEventHandler(handler EventHandler) {
 	d.eventHandlers = append(d.eventHandlers, handler)
 }
 
-func (d *discoverer) addService(obj interface{}) {
-	svc := obj.(*corev1.Service)
-	d.updateServiceCIDR(svc)
-}
-
-func (d *discoverer) updateService(_, obj interface{}) {
+func (d *Discoverer) addService(obj interface{}) {
 	svc := obj.(*corev1.Service)
-	d.updateServiceCIDR(svc)
+	klog.V(2).InfoS("Processing Service ADD event", "Service", klog.KObj(svc))
+	d.queue.Add(types.NamespacedName{Namespace: svc.Namespace, Name: svc.Name})
 }
 
-func (d *discoverer) updateServiceCIDR(svc *corev1.Service) {
-	clusterIPs := svc.Spec.ClusterIPs
-	if len(clusterIPs) == 0 {
-		return
+func (d *Discoverer) updateService(old, obj interface{}) {
+	oldSvc := old.(*corev1.Service)
+	curSvc := obj.(*corev1.Service)
+	klog.V(2).InfoS("Processing Service UPDATE event", "Service", klog.KObj(curSvc))
+	if !slices.Equal(oldSvc.Spec.ClusterIPs, curSvc.Spec.ClusterIPs) {
+		d.queue.Add(types.NamespacedName{Namespace: curSvc.Namespace, Name: curSvc.Name})
 	}
+}
 
+func (d *Discoverer) updateServiceCIDR(svcs ...*corev1.Service) {
```

Comment:
It's not clear to me that the current implementation of `updateServiceCIDR` will always be correct if multiple workers / goroutines can call `updateServiceCIDR` simultaneously?
That being said, it should not matter given that you start a single worker goroutine in `Run`.

Reply:
It may not work if it's called simultaneously. There is no need to have multiple workers for this task as it just checks if an IP is in a subnet in most cases (CIDR expansion can only happen few times).